README.md
# Final h-encore

Push the h-encore exploit for PS VITA and PS TV through a Windows, macOS or Linux GUI.

## Credits

see [CREDITS.md](CREDITS.md)

## Common usage

1. Download a pre-built executable binary below.
1. If you want to update PS Vita to firmware 3.60, 3.65 or 3.68 through USB connection, download related Update Packages [here](https://darthsternie.net/index.php/ps-vita-firmwares/) and extract the `PSP2UPDAT.PUP` to the same folder as this tool
1. If you want to install VitaShell or enso by transferring in Content Manager, download prebuilt zip from `releases` tab and put in the same folder of this tool
1. Connect your PS Vita to your computer via USB
1. Run the executable and follow the on-screen directions

### Prebuilt binaries

Download a pre-built executable binary below and follow instructions. Supported firmwares: 3.60, 3.61, 3.65-3.72

- For Windows get the [latest release](https://github.com/soarqin/finalhe/releases/latest)
  - if you have not installed USB driver for PS Vita: install `QcmaDriver_winusb.exe` (also on the releases page) 
- For macOS, the last pre-built [release is v1.5](https://github.com/soarqin/finalhe/releases/tag/v1.5)
- For linux openSUSE
  1. add a home [repository](https://software.opensuse.org/package/finalhe) to your local software repositories:
      - `sudo addrepo -f http://download.opensuse.org/repositories/home:/seilerphilipp/openSUSE_Leap_15.0/`
      - note: if using Leap 42.3, replace the version in the url with `42.3`
  2. install package
      - `sudo zypper install finalhe`
  3. Run "FinalHE"
      - in your terminal type `FinalHE`
- For Arch Linux
  1. Install the [`finalhe-git`](https://aur.archlinux.org/packages/finalhe-git/) package from AUR:
      - `yay -S finalhe-git`
  2. Run `finalhe` in your terminal

## Build from source

### Prerequisites 

1. macOS: install [brew](https://brew.sh), build tools (`xcode-select --install`) then install dyld, libusb, pkg-config & qt5 through brew (`brew install dyld-headers libusb pkg-config qt5`)
1. Linux:
   - Debian/Ubuntu: install build-essential, libxml2-dev, libusb-dev, libusb-1.0-0-dev, zlib-dev or zlib1g-dev, qtbase5-dev, qttools5, cmake(if use cmake to build)
   - Fedora/CentOS: group install "Development Tools", install libxml2-devel, libusb-devel, zlib-devel, qt5-qtbase-devel, qt5-qtbase, qt5-linguist, cmake3(if use cmake to build)
   - openSUSE: install cmake >= 11.0, gcc-c++, zlib-devel, libxml2-devel, libQt5Widgets-devel, libQt5Network-devel, libqt5-linguist-devel, libusb-compat-devel
   - Arch: install base-devel, libxml2, libusb, zlib, qt5, cmake (if using cmake to build)

### Build from source

You can choose either `qmake` or `cmake` to build

- cmake: run `cmake` to generate Makefile for compiling
  - macOS: it cannot produce app bundle, and you need to specify `CMAKE_PREFIX_PATH` if Qt is not installed in default location: `cmake -DCMAKE_PREFIX_PATH=<Path of Qt Root> <Path of Project Root>`. If you have installed qt5 with brew, and are on project root directory, you can run with `cmake -DCMAKE_PREFIX_PATH=/usr/local/opt/qt@5 .`. Then run `make` and you can open it by double clicking at `src/FinalHE`.
- qmake: run `qmake` to generate Makefile for compiling, run `make lcopy` in `src` folder to compile translations and copy them to binary folder

## Contribute translations

- For coders using Qt:
   1. If using qmake, add your language to this line in src/src.pro: `TRANSLATIONS += ...`, and re-generate Makefile from qmake, then run `make lupdate` in `src` folder, you will get new generated .ts files in src/languages
   1. If  using cmake, add your language to this line in src/CMakeLists.txt: `set(TRANSLATION_FILES ...`, and re-generate Makefile/project from cmake, compile it, you will get new generated .ts files in src/languages
   1. Open .ts files with Qt Linquist tool and translate strings into native language
   1. If using qmake, run `make lcopy` in src folder and you will get compiled .qm files in languages folder
   1. If using cmake, compile the project and you will get compiled .qm files in language folder
- For non-coders:
   1. Copy translations/en_US.ts to a new file with filename in [IETF language tag](https://datahub.io/core/language-codes/r/3.html) form (but replace "-" with "_")
   1. Open the .ts file, it is just in xml format, go through all elements of `<translation type="unfinished">` to translate, and remove the property `type="unfinished"` from translated items


CHANGES
v1.93
1. (#125) Added support for 3.74 with existing h-encore^2
2. (#108) Project builds on Apple M1 now

v1.92
1. Added support for 3.73 with existing h-encore^2

v1.91
1. Added support for 3.72 with updated h-encore²

v1.9
1. Support 3.69-3.71 with h-encore²
2. Fixed a bug which causes memcore not working on 3.61

v1.81
1. Remove deflate from h-encore to get payload loading slight faster
2. Support 3.61/3.67 now, you won't be asked to update on these firmwares (#46)
3. Reduced C2-12821-1 error (#47)
4. Fixed mismatch firmware versions display on firmware update

v1.8
1. Support for memecore(h-encore 2.0 for VITA 3.60 support), compiled from [my fork](https://github.com/soarqin/h-encore), thanks to [CelesteBlue-dev](https://github.com/CelesteBlue-dev) for his port
2. Can trim h-encore app to only 6.5MB now, thanks to [KuromeSan](https://github.com/KuromeSan)
3. Fixed a bug that button is reset to be clickable when another info sync from PSVITA is received
4. Fixed French language

v1.71
1. Updated official fix for h-encore 2.0

v1.7
1. Updated h-encore to 2.0

v1.6
1. Calculate Backup key from AID offline (#43), thanks to [KuromeSan's code](https://bitbucket.org/SilicaAndPina/cmakeyderiv)

v1.5
1. New feature: Check firmware version and support firmware update 3.65/3.68 through USB connection
2. New feature: build prepared applications for transferring through Content Manager
3. Added translation: Russian

v1.4
1. Embedded h-encore.zip into binary
2. Added translation: Japanese

v1.3
1. Added translations: Czech, Italian, Spanish & Turkish
2. Fixed crash on downloading h-encore.zip
3. Made prebuilt OS X binary compatible with Mavericks 10.9+
4. Other minor fixes

v1.2
1. Supported WiFi connection for Content Manager
2. Ported to OS X/macOS and Linux

v1.1
1. Fixed incorrect installation process

v1.0
1. Finished all basic functions


src/finalhe.cc
/*
 *  Final h-encore, Copyright (C) 2018  Soar Qin
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "finalhe.hh"

#include "package.hh"
#include "vita.hh"
#include "version.hh"

#include <QLocale>
#include <QDir>
#include <QMessageBox>
#include <QStandardPaths>
#include <QDebug>
#include <QSettings>

FinalHE::FinalHE(QWidget *parent): QMainWindow(parent) {
    ui.setupUi(this);
    setWindowTitle("Final h-encore v" FINALHE_VERSION_STR);
    setWindowIcon(QIcon(":/main/resources/images/finalhe.png"));
    setFixedSize(600, 400);
    ui.centralWidget->setFixedSize(600, 400);
    ui.extraItems->setFixedSize(200, 400);
    ui.extraItems->hide();
    ui.expandButton->setArrowType(Qt::RightArrow);
    ui.expandButton->setFixedWidth(12);

    ui.textMTP->setStyleSheet("QLabel { color : blue; }");

    ui.progressBar->setMaximum(100);
    QDir baseDir, dir(qApp->applicationDirPath());
#ifdef __APPLE__
    dir.cdUp(); dir.cdUp(); dir.cdUp();
#endif
#ifdef _WIN32
    baseDir = dir;
    if (!baseDir.mkpath("data") || !(baseDir.cd("data"), baseDir.exists()))
#endif
    {
        baseDir = QStandardPaths::writableLocation(QStandardPaths::AppDataLocation);
        baseDir.mkpath("data");
        baseDir.cd("data");
    }
    vita = new VitaConn(baseDir.path(), dir.path(), this);
    pkg = new Package(baseDir.path(), dir.path(), this);
    int useSysLang = 0;
    QSettings settings;
    QString langLoad = settings.value("language").toString();
    ui.comboLang->addItem(trans.isEmpty() ? "English" : trans.translate("base", "English"));
    dir = qApp->applicationDirPath();
    if (
#if defined(__APPLE__)
        dir.cdUp() &&
#elif defined(__unix__)
        dir = QStandardPaths::writableLocation(QStandardPaths::AppDataLocation),
#endif
        dir.cd("language")) {
        QStringList ll = dir.entryList({"*.qm"}, QDir::Filter::Files, QDir::SortFlag::IgnoreCase);
        for (auto &p: ll) {
            QString compPath = dir.filePath(p);
            QTranslator transl;
            if (transl.load(compPath)) {
                int index = ui.comboLang->count();
                ui.comboLang->addItem(transl.translate("base", "LANGUAGE"), compPath);
                if (useSysLang != 0) continue;
                if (langLoad.isNull()) {
                    if (QFileInfo(p).baseName() == QLocale::system().name()) {
                        useSysLang = index;
                    }
                } else if (compPath == langLoad) {
                    useSysLang = index;
                }
            }
        }
    }

    pkg->setTrim(ui.checkTrim->checkState() == Qt::Checked);


    connect(ui.btnStart, SIGNAL(clicked()), this, SLOT(onStart()));
	connect(ui.comboLang, SIGNAL(currentIndexChanged(int)), this, SLOT(langChange()));
    connect(ui.checkTrim, SIGNAL(stateChanged(int)), this, SLOT(trimState(int)));
    connect(ui.expandButton, SIGNAL(clicked()), this, SLOT(toggleExpanding()));
    connect(ui.extraItems, SIGNAL(itemClicked(QListWidgetItem*)), this, SLOT(extraItemsChanged(QListWidgetItem*)));
    connect(vita, &VitaConn::gotAccountId, this, [this](QString accountId) {
        pkg->calcBackupKey(accountId);
        enableStart();
    });
    connect(vita, SIGNAL(receivedPin(QString, int)), this, SLOT(displayPin(QString, int)));
    connect(vita, SIGNAL(completedPin()), this, SLOT(clearPin()));
    connect(pkg, SIGNAL(createdPsvImgs()), vita, SLOT(buildData()));
    connect(vita, SIGNAL(builtData()), pkg, SLOT(finishBuildData()));

    connect(vita, SIGNAL(setStatusText(QString)), this, SLOT(setTextMTP(QString)));
    connect(pkg, SIGNAL(setStatusText(QString)), this, SLOT(setTextPkg(QString)));
    connect(pkg, SIGNAL(setPercent(int)), ui.progressBar, SLOT(setValue(int)));

    vita->process();
    pkg->tips();

    if (useSysLang) {
        ui.comboLang->setCurrentIndex(useSysLang);
        langChange();
    }
    updateExpandArea();
}

FinalHE::~FinalHE() {
    delete pkg;
    delete vita;
}

void FinalHE::langChange() {
    QVariant var = ui.comboLang->currentData();
    QString compPath = var.toString();
    loadLanguage(compPath);
    pkg->tips();
    vita->updateStatus();
    QSettings().setValue("language", compPath);
}

void FinalHE::onStart() {
    ui.btnStart->setEnabled(false);
    emit pkg->startDownload();
}

void FinalHE::updateExpandArea() {
    ui.extraItems->clear();
    bool hasFirmware = false;
    for (int i = 0; i < 3; ++i) {
        bool has = false;
        switch(i) {
            case 0:
                has = vita->has360Update(); break;
            case 1:
                has = vita->has365Update(); break;
            case 2:
                has = vita->has368Update(); break;
            default: break;
        }
        if (has) {
            auto *item = new QListWidgetItem(tr("Firmware %1").arg(i == 0 ? "3.60" : (i == 1 ? "3.65" : "3.68")));
            item->setFlags(Qt::ItemIsEnabled | Qt::ItemIsUserCheckable);
            if (!hasFirmware) {
                auto *item0 = new QListWidgetItem(tr("-- Firmware update --"));
                item0->setFlags(Qt::NoItemFlags);
                ui.extraItems->addItem(item0);
                hasFirmware = true;
            } else
                item->setCheckState(Qt::Unchecked);
            item->setData(Qt::UserRole, i + 1);
            ui.extraItems->addItem(item);
        }
    }
    const auto &extraApps = pkg->getExtraApps();
    bool hasExtra = false;
    for (auto &p : extraApps) {
        if (!hasExtra) {
            auto *item0 = new QListWidgetItem(tr("-- Additional applications --"));
            item0->setFlags(Qt::NoItemFlags);
            ui.extraItems->addItem(item0);
            hasExtra = true;
        }
        auto *item = new QListWidgetItem(p.name);
        item->setFlags(Qt::ItemIsEnabled | Qt::ItemIsUserCheckable);
        ui.extraItems->addItem(item);
        item->setData(Qt::UserRole, p.titleId);
        bool defsel = p.titleId == "VITASHELL";
        item->setCheckState(defsel ? Qt::Checked : Qt::Unchecked);
        if (defsel) pkg->selectExtraApp(p.titleId, true);
    }
    if (hasFirmware || hasExtra) {
#ifndef _WIN32
        ui.centralWidget->setFixedSize(588, 400);
#endif
        ui.expandButton->show();
    } else {
        setFixedSize(600, 400);
        ui.centralWidget->setFixedSize(600, 400);
        ui.expandButton->hide();
        ui.extraItems->hide();
    }
}

bool FinalHE::checkFwUpdate() {
    if (vita->getDeviceVersion().isEmpty()) return true;
    if (vita->getDeviceVersion() < "3.60") {
        ui.textPkg->setText(tr("Fimrware version is not supported by h-encore.") + "\n"
            + tr("Update to %1 first.").arg("3.60/3.65/3.68") + "\n"
            + ((vita->has360Update() || vita->has365Update() || vita->has368Update())
            ? tr("On PS Vita:\nSettings -> System Update -> Update by Connecting to a PC")
            : tr("To update through USB:\nPut Update Package(.PUP) in this tool's folder and restart the tool")));
        return false;
    } else if (vita->getDeviceVersion() < "3.65") {
        bool needUpdate = vita->getDeviceVersion() > "3.61";
        int count = ui.extraItems->count();
        for (int i = 0; i < count; ++i) {
            auto *eitem = ui.extraItems->item(i);
            if (eitem->flags() == Qt::NoItemFlags) continue;
            bool ok;
            int n = eitem->data(Qt::UserRole).toInt(&ok);
            if (ok) {
                if (n == 1) {
                    ui.extraItems->removeItemWidget(eitem);
                    delete eitem;
                    count = ui.extraItems->count();
                    --i;
                } else if (needUpdate) {
                    eitem->setCheckState(n > 1 ? Qt::Checked : Qt::Unchecked);
                }
            }
        }
        if (needUpdate) {
            ui.textPkg->setText(tr("Fimrware version is not supported by h-encore.") + "\n"
                + tr("Update to %1 first.").arg("3.65/3.68") + "\n"
                + (vita->has368Update()
                ? tr("On PS Vita:\nSettings -> System Update -> Update by Connecting to a PC")
                : tr("To update through USB:\nPut Update Package(.PUP) in this tool's folder and restart the tool")));
            return false;
        }
    } else if (vita->getDeviceVersion() < "3.68") {
        int count = ui.extraItems->count();
        for (int i = 0; i < count; ++i) {
            auto *eitem = ui.extraItems->item(i);
            if (eitem->flags() == Qt::NoItemFlags) continue;
            bool ok;
            int n = eitem->data(Qt::UserRole).toInt(&ok);
            if (ok) {
                if (n == 1 || n == 2) {
                    ui.extraItems->removeItemWidget(eitem);
                    delete eitem;
                    count = ui.extraItems->count();
                    --i;
                }
            }
        }
    }
    return true;
}

void FinalHE::enableStart() {
    if (checkFwUpdate()) {
        if (vita->getDeviceVersion() <= "3.61")
            pkg->setCoreType(ECoreType::Memecore);
        else if (vita->getDeviceVersion() <= "3.68")
            pkg->setCoreType(ECoreType::HEncore);
        else
            pkg->setCoreType(ECoreType::HEncore2);
        ui.textPkg->setText(tr("Click button to START!"));
        ui.btnStart->setEnabled(true);
    }
}

void FinalHE::trimState(int state) {
    pkg->setTrim(state == Qt::Checked);
}

void FinalHE::setTextMTP(QString txt) {
    ui.textMTP->setText(txt);
}

void FinalHE::setTextPkg(QString txt) {
    ui.textPkg->setText(txt);
}

void FinalHE::displayPin(QString onlineId, int pin) {
    ui.textPkg->setText(tr("Registering device: %1\nInput this PIN on PS Vita: %2").arg(onlineId).arg(pin, 8, 10, QChar('0')));
}

void FinalHE::clearPin() {
    ui.textPkg->setText(tr("Registered device."));
}

void FinalHE::toggleExpanding() {
    if (expanding) {
        setFixedSize(600, 400);
        ui.extraItems->hide();
        ui.expandButton->setArrowType(Qt::RightArrow);
        expanding = false;
    } else {
#ifdef _WIN32
        setFixedSize(812, 400);
#else
        setFixedSize(800, 400);
#endif
        ui.extraItems->show();
        ui.expandButton->setArrowType(Qt::LeftArrow);
        expanding = true;
    }
}

void FinalHE::extraItemsChanged(QListWidgetItem *item) {
    QVariant var = item->data(Qt::UserRole);
    bool ok;
    int fwidx = var.toInt(&ok);
    bool checked = item->checkState() == Qt::Checked;
    if (!ok) {
        pkg->selectExtraApp(var.toString(), checked);
    } else {
        if (checked) {
            switch (fwidx) {
            case 1:
                vita->setUse360Update();
                break;
            case 2:
                vita->setUse365Update();
                break;
            case 3:
                vita->setUse368Update();
                break;
            default:
                vita->setUseNoUpdate();
                break;
            }
            int count = ui.extraItems->count();
            for (int i = 0; i < count; ++i) {
                auto *eitem = ui.extraItems->item(i);
                if (eitem->flags() == Qt::NoItemFlags || eitem == item) continue;
                bool ok;
                int n = eitem->data(Qt::UserRole).toInt(&ok);
                if (ok && n > 0) {
                    eitem->setCheckState(Qt::Unchecked);
                }
            }
        } else
            vita->setUseNoUpdate();
    }
}

void FinalHE::loadLanguage(const QString &s) {
    qApp->removeTranslator(&trans);
    if (trans.load(s)) {
        qApp->installTranslator(&trans);
    }
    ui.retranslateUi(this);
    ui.comboLang->setItemText(0, trans.isEmpty() ? "English" : trans.translate("base", "English"));
    updateExpandArea();
    checkFwUpdate();
}
